### PR TITLE
Crew objectives

### PR DESCRIPTION
**Pull request by Carbonhell**
changes: 
- rscadd: "Adds crew objectives. When joining, all jobs excluding silicons and antags will get an objective related to their job to accomplish. At roundend, the successful crewmembers are listed along, and the department with the most successful crewmembers is also listed. Obviously, those objectives do NOT make you an antagonist."

Big thanks to: Badmex, Shamzie, Mini_Meatwad and all the other lads who helped me in my several tests ayy lmao

### CODE REVIEW

File: code/controllers/subsystem/ticker.dm [hunk 1]
```diff
@@ -428,6 +428,40 @@ var/datum/subsystem/ticker/ticker
...
+		var/mob/M = A
+		if(M.mind && M.mind.objectives)
+			if(locate(/datum/objective/crew) in M.mind.objectives)
+				var/youdidit = TRUE
+				for(var/datum/objective/crew/C in M.mind.objectives)
+					if(!C.check_completion())
+						youdidit = FALSE
+				if(youdidit)
+					successfulcrew += M.name
+	world << "<font color='green'><b>The following crewmembers managed to complete all their crew objectives: [jointext(successfulcrew, ", ")]</b></font>"
```
Comment (ghost) [inline code/controllers/subsystem/ticker.dm]: A massive paragraph in solid green is going to look awful.
Could we change that it's just black bold text and use  <span class='greenannounce'></span>for all the successful crewmembers please?

**Comment by Carbonhell** [comment]
done baw

**Comment by ghost** [comment]
I am pleased.
+1

**Comment by Kierany9** [comment]
-1 terrible idea

**Comment by Spacedong** [comment]
please specify that they are strictly optional so we dont turn into autist station 13

**Comment by ghost** [comment]
Daily reminder that this was my suggestion

I did like 60% of the work, carbon just filled in the details of actually making it happen

I accept royalties in antag tokens only

**Comment by Kokojos** [comment]
Definitively you must change the rules to make sure that the ''You're not an antag if you don't have objectives'' rule dosen't confound stupid players.

Also, spacedong, yes they are optionnal, like your life. Honk honk.

+1/3rd this magnificient piece of shit.

**Comment by Badmex** [comment]
Yeah +1/3 The testing was already funny.

File: code/game/gamemodes/objective.dm [hunk 2]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+/datum/objective/crew/medical/clone/check_completion()
+	var/datum/feedback_variable/medicalfeedback = blackbox.find_feedback_datum("medicalshit")
+	var/peepscloned = medicalfeedback.get_value()
+	if(peepscloned >= peepstoclone)
+		..(1)
+		return 1
+
+/datum/objective/crew/medical/clone/update_explanation_text()
+	..()
+	explanation_text = "Clone atleast [peepstoclone] people."
```
Comment (Kokojos) [inline code/game/gamemodes/objective.dm]: at least*

File: code/game/gamemodes/objective.dm [hunk 3]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+		else
+			done = FALSE
+	if(done)
+		..(1)
+		return 1
+
+/datum/objective/crew/medical/makechem/update_explanation_text()
+	..()
+	var/list/chemlist = splittext(chemid, "#")
+	explanation_text = "Make atleast [round(amount/chemlist.len)] units of each chemical: [chemname]"
```
Comment (Kokojos) [inline code/game/gamemodes/objective.dm]: same here.

**Comment by McHatters** [comment]
Yes please +1

**Comment by ghost** [comment]
I agree that rule 6 should be expanded upon but if people are really going to use rule 6 as their rule to grief because "I have objective to keep bar clean so I kill everyone else so they don't make mess" or some dumb shit then I'd be shock if admins are hesitant to ban.

Not only that, if someone sees that they got objectives and thinks that means they can grief then they probably didn't read the rules in the first place.

File: code/game/gamemodes/objective.dm [hunk 4]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+	for(var/turf/simulated/floor/plasteel/shuttle/red/R in area_contents(E))
+		for(var/mob/living/carbon/human/H in R)
+			if(H.mind)
+				if(H.mind.special_role)
+					..(1)
+					return 1
+
+/datum/objective/crew/security/arrestantag/update_explanation_text()
+	..()
+	explanation_text = "Have atleast one human antagonist in the shuttle brig when the shift ends."
```
Comment (Kokojos) [inline code/game/gamemodes/objective.dm]: atleast!

Comment (ghost) [inline code/game/gamemodes/objective.dm]: Holy _FUCK_ carbon! Atleast isn't a word!

File: code/game/gamemodes/objective.dm [hunk 5]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+	var/list/details = details2list("ore_mined", " ", "|")
+	var/textedpath = "[oretype]"
+	if(textedpath in details)
+		if(details[textedpath] >= oreamt)
+			..(1)
+			return 1
+
+/datum/objective/crew/supply/mineore/update_explanation_text()
+	..()
+	explanation_text = "Mine atleast [oreamt] [initial(oretype.name)]s."
```
Comment (Kokojos) [inline code/game/gamemodes/objective.dm]: Atrest

File: code/game/gamemodes/objective.dm [hunk 6]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+
+/datum/objective/crew/research/getslime/check_completion()
+	for(var/mob/living/simple_animal/slime/S in mob_list)
+		if(S.colour == colorneeded)
+			..(1)
+			return 1
+
+/datum/objective/crew/research/getslime/update_explanation_text()
+	..()
+	explanation_text = "Produce atleast a single [colorneeded] slime."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 7]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+	for(var/obj/machinery/bot/B in craftedbots)
+		if(istype(B, bottype))
+			botamt++
+	if(botamt >= amt)
+		..(1)
+		return 1
+
+/datum/objective/crew/research/bots/update_explanation_text()
+	..()
+	explanation_text = "Craft atleast [amt] number of [initial(bottype.name)]s and ensure they survive till the shift ends. Roundstart ones don't count, so name them!"
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 8]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+		droneamt++
+	for(var/mob/living/simple_animal/drone/D in living_mob_list)
+		droneamt++
+	if(droneamt >= amt)
+		..(1)
+		return 1
+
+/datum/objective/crew/research/drones/update_explanation_text()
+	..()
+	explanation_text = "Build atleast [amt] drones, doesn't matter if active or not, but they must be kept intact."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 9]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+	for(var/mob/living/silicon/robot/R in living_mob_list)
+		if(R.stat != DEAD)
+			amt--
+	if(!amt)
+		..(1)
+		return 1
+
+/datum/objective/crew/research/borgs/update_explanation_text()
+	..()
+	explanation_text = "Have atleast [amt] alive cyborgs at the end of the shift."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 10]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+			if(Player.stat != DEAD && !isbrain(Player))
+				num_survivors++
+	var/percentage = round((num_survivors/joined_player_list.len)*100, 0.1)
+	if(percentage >= percneeded*10)
+		..(1)
+		return 1
+
+/datum/objective/crew/medical/survivor/update_explanation_text()
+	..()
+	explanation_text = "Reach atleast the [percneeded*10]% of survivors at shift end."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 11]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+		return
+	var/value = fbdetails[owner.key]
+	if(value)
+		if(value >= peepsneeded)
+			..(1)
+			return 1
+
+/datum/objective/crew/medical/defib/update_explanation_text()
+	..()
+	explanation_text = "Successfully defib atleast [peepsneeded] people."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 12]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+			explanation_text += " and " //since it's either 1 or 2.
+	explanation_text += " are completely clean when the shift ends. Please ask for permission before entering in certain places!"
+
+//CIVILIAN! holy shit i didn't believe i'd reach this
+/datum/objective/crew/civilian
+	mytag = "Civilian"
+	jobs = "Captain#Librarian#Lawyer#Chaplain#Clown#Mime#Assistant"
+
+//keep dat fukken disk secure
+/datum/objective/crew/civilian/datfukkendisk
```
Comment (Kokojos) [inline code/game/gamemodes/objective.dm:1755]: Why isin't this Command instead of civilian?

File: code/game/gamemodes/objective.dm [hunk 13]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+	points = rand(1000,2000)
+
+/datum/objective/crew/supply/points/check_completion()
+	if(SSshuttle.points >= points)
+		..(1)
+		return 1
+
+/datum/objective/crew/supply/points/update_explanation_text()
+	..()
+	explanation_text = "Have atleast [points] supply points at shift end. You can achieve this by shipping plasma, exotic plants, syndicate secret documents or research saved on disks."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 14]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+/datum/objective/crew/supply/miningpoints/check_completion()
+	if(owner && owner.current)
+		for(var/obj/item/weapon/card/id/I in owner.current)
+			if(I.mining_points >= points)
+				..(1)
+				return 1
+
+/datum/objective/crew/supply/miningpoints/update_explanation_text()
+	..()
+	explanation_text = "Have atleast [points] mining points on any ID card in your inventory at shift end."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 15]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+					plants[path]--
+		for(var/i in plants)
+			if(plants[i])
+				return 0
+		..(1)
+		return 1
+
+/datum/objective/crew/service/collectplants/update_explanation_text()
+	..()
+	explanation_text = "Collect atleast "
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 16]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+	for(var/i in dat)
+		if(dat[i] == "/obj/item/device/pda/clown")
+			amt--
+	if(!amt)
+		..(1)
+		return 1
+
+/datum/objective/crew/civilian/slip/update_explanation_text()
+	..()
+	explanation_text = "Slip the crew atleast [amt] times with your PDA."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

File: code/game/gamemodes/objective.dm [hunk 17]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+				for(var/datum/newscaster/feed_message/message in F.messages)
+					if(!(message in articleschecked))
+						if(findtext(F.author, owner.current.job) && length(message.body) >= char)
+							articleschecked += message
+							..(1)
+							return 1
+
+/datum/objective/crew/civilian/reporter/update_explanation_text()
+	..()
+	explanation_text = "Make atleast [articles] articles, they must be atleast [char] characters long."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: fuck

**Comment by Spacedong** [comment]
does this specifically say when you get them that they're optional and you shouldn't play like an antag to get your objectives done, if it doesn't, then add it

File: code/game/gamemodes/objective.dm (hunk 12 above)
Comment (Carbonhell) [inline code/game/gamemodes/objective.dm:1755]: cause every head was in his own department, but making one department just for captain felt retarded so i moved him to civilian

File: code/controllers/subsystem/jobs.dm [hunk 18]
```diff
@@ -447,3 +455,63 @@ var/datum/subsystem/job/SSjob
...
+	else
+		return
+
+/datum/subsystem/job/proc/job2obj(rank, blacklist)
+	var/list/objs = list()
+	var/list/acceptableobjs = list()
+	var/dept = job2dept(rank)
+	for(var/A in departments)
+		if(departments[A] == dept)
+			objs = typesof(A) - A
```
Comment (JohnGinnane) [inline code/controllers/subsystem/jobs.dm:483]: Should we break after this is found?

**Comment by Carbonhell** [comment]
uhh no spacedong i don't want people to ignore them

**Comment by Carbonhell** [comment]
grammar fixed

**Comment by GenericGoose** [comment]
WE GOON NOW?
i want this
economy with money next?

**Comment by Kokojos** [comment]
Perfect code here. Codecheck pass.

**Comment by Nebyab** [comment]
Once the rules are updated to reflect this, then and only then can we merge.

**Comment by myontasm** [comment]
good stuff

**Comment by Spacedong** [comment]
@generalgoose I want comfy barber shop and arcade/VR rooms next, then we can do economy

**Comment by GenericGoose** [comment]
@Spacedong  same
p sure tg has vr already

**Comment by ghost** [comment]
VR more like virginal rage.

Reee.

**Comment by GenericGoose** [comment]
vr wgw erp when

**Comment by SplergeBucket** [comment]
Rules would have to be amended if this goes in.
It'd provide something for crewmembers to do which I like.

File: code/controllers/subsystem/jobs.dm (hunk 18 above)
Comment (Carbonhell) [inline code/controllers/subsystem/jobs.dm:483]: added

File: code/game/gamemodes/objective.dm [hunk 19]
```diff
@@ -872,3 +879,1060 @@ var/global/list/possible_items_special = list()
...
+		if(powerrequired <= 0)
+			powerrequired = 0
+			break
+	if(!powerrequired)
+		..(1)
+		return 1
+
+/datum/objective/crew/engineering/fullsmes/update_explanation_text()
+	..()
+	explanation_text = "Get the station's SMESes to a complessive charge of [powerrequired] millions of Watts. Keep in mind that each SMES can hold 3,3 millions of Watts."
```
Comment (ghost) [inline code/game/gamemodes/objective.dm]: 3,3 is Latin style. It'd probably make more sense to write it like 3.3

Comment (JohnGinnane) [inline code/game/gamemodes/objective.dm]: yes please)

**Comment by ghost** [comment]
Rules have been updated.
Merging.

**Comment by GenericGoose** [comment]
nice meme!

**Comment by Carbonhell** [comment]
thanks fam!

**Pull request merged**